### DISCUSSION TITLE
Exclude services based on the environment

### DISCUSSION THREAD

**Discussion by ker0x** [symfony/symfony#50584] [Q&A] [answered]
TLDR: 

> I want my `LocalAdapter` to be excluded from the list of adapters in the `test` environment, which doesn't seem to be the case because my application seems to use it anyway as my tests fail to check a wrong number.

---

Hi,

I'm having a little trouble understanding how the `#[When]` attribute works.

I have a manager who checks the validity of a legal unit against a number for a given country using adapters.
In my `dev` environment, I want it to always return `true` for simplicity's sake, so I created a `LocalAdapter`.

Suppose I had an interface like this one:

```php
#[AutoconfigureTag(name: 'app.legal.adapter')]
interface AdapterInterface
{
    public static function getCountry(): string;

    public function isValidLegalUnit(string $registrationNumber): bool;
}
```

And 2 adapters that implement this interface:

```php
final class InseeAdapter implements AdapterInterface
{
    private const COUNTRY = 'FR';

    public static function getCountry(): string
    {
        return self::COUNTRY;
    }

    public function isValidLegalUnit(string $registrationNumber): bool
    {
        // Call to an external API to check registration number validity
    }
}

```

```php
#[When('dev')]
final class LocalAdapter implements AdapterInterface
{
    private const COUNTRY = 'FR';

    public static function getCountry(): string
    {
        return self::COUNTRY;
    }

    public function isValidLegalUnit(string $registrationNumber): bool
    {
        return true;
    }
}
```

These adapters are then retrieved in a handler with a `TaggedLocator` attribute:

```php
final readonly class LegalManager implements LegalManagerInterface
{
    public function __construct(
        #[TaggedLocator(tag: 'app.legal.adapter', defaultIndexMethod: 'getCountry')]
        private ServiceLocator $adapters,
    ) {
    }

    public function isValidLegalUnit(string $registrationNumber, string $country): bool
    {
        return $this->getAdapter($country)->isValidLegalUnit($registrationNumber);
    }

    private function getAdapter(string $country): AdapterInterface
    {
        if (!$this->adapters->has($country)) {
            throw new UnsupportedCountryException(sprintf('Country "%s" is not yet supported.', $country));
        }

        /** @var AdapterInterface $adapter */
        $adapter = $this->adapters->get($country);

        return $adapter;
    }
}
```

When I run `bin/console debug:container --tag app.legal.adapter`, i get this:

```bash
Symfony Container Services Tagged with "app.legal.adapter" Tag
==============================================================

 ----------------------------------------------- -----------------------------------------------
  Service ID                                      Class name
 ----------------------------------------------- -----------------------------------------------
  App\Legal\Infrastructure\Adapter\InseeAdapter   App\Legal\Infrastructure\Adapter\InseeAdapter
  App\Legal\Infrastructure\Adapter\LocalAdapter   App\Legal\Infrastructure\Adapter\LocalAdapter
 ----------------------------------------------- -----------------------------------------------
```

which is totally OK as the command is running in the `dev` environment by default.

However, if I run the same command with `--env=test`, I get the same result:

```
bin/console debug:container --env=test --tag app.legal.adapter
```

```bash
Symfony Container Services Tagged with "app.legal.adapter" Tag
==============================================================

 ----------------------------------------------- -----------------------------------------------
  Service ID                                      Class name
 ----------------------------------------------- -----------------------------------------------
  App\Legal\Infrastructure\Adapter\InseeAdapter   App\Legal\Infrastructure\Adapter\InseeAdapter
  App\Legal\Infrastructure\Adapter\LocalAdapter   App\Legal\Infrastructure\Adapter\LocalAdapter
 ----------------------------------------------- -----------------------------------------------
```

Given the use of the `#[When('dev')]` attribute on `LocalAdapter`, shouldn't it be excluded from the result in the test environment?

I'm using Symfony `6.3`!

**Comment by ghost** [comment]
Yes it should be excluded. Try to reproduce it on symfony/skeleton.

**Comment by stof** [comment]
My guess is that the change done in https://github.com/symfony/symfony/pull/48606 was incomplete

**Comment by MatTheCat** [comment]
I cannot reproduce your issue using Symfony 6.3 :thinking: 

Could you provide a [reproducer](https://symfony.com/doc/current/contributing/code/reproducer.html)?

**Reply by ker0x** [reply]
On a basic Symfony structure, this works. 

However, my project is build around DDD, with the interface being in `App/Legal/Domain/Adapter` and the classes in `App/Legal/Infrastructure/Adapter` with each Domain having it's own service definition.

Could a faulty configuration be the cause of my problem?

To get a better idea of how my code is structured, you can take a look at this repository: [apip-ddd](https://github.com/mtarld/apip-ddd)

**Reply by ker0x** [reply]
Here is a reproducer: https://github.com/ker0x/reproducer-discussions-50584

No need for the whole docker thing, just `composer install` and running the console command is enougth

**Reply by MatTheCat** [reply, marked as answer]
You’re using the validator’s `When` attribute, not the dependency injection’s :sweat_smile: 

```diff
- use Symfony\Component\Validator\Constraints\When;
+ use Symfony\Component\DependencyInjection\Attribute\When;
```

**Reply by ghost** [reply]
haha what a fail :) this is this kind of developer's failure where you regret the time spent on debugging the most

**Reply by ker0x** [reply]
Oh, my God! What an idiot I am 🤦 !

Sorry folks for the wasted time due to my stupidity...

**Reply by MatTheCat** [reply]
Well in fact I made the exact same mistake while trying to reproduce the issue, because I didn’t know about [the `When` constraint](https://symfony.com/doc/current/reference/constraints/When.html), so I let my IDE autocomplete this attribute as it came first. Since my IDE also collapses `use`s  by default, I didn’t notice right away..!

**Reply by ker0x** [reply]
Welcome to the club!

I also didn't know about it until today, and I'm finally discovering a use case for it in my project. 😅
